### PR TITLE
 Reduced muon validation for PhaseII Upgrade high PU scenarios (81X)

### PR DESCRIPTION
**Pull request by calabria**
@kpedro88

### CODE REVIEW

**Comment by slava77** [comment]
can you also fix the timing problem in https://github.com/cms-sw/cmssw/blob/CMSSW_8_1_0_pre16/SimMuon/MCTruth/src/MuonAssociatorByHitsHelper.cc#L1192
the suggestion was to define ``` tbegin = trpart->g4Track_begin(); ``` and tend in the beginning of the function scope instead of triple-nested for loop.

**Comment by calabria** [comment]
@slava77 @abbiendi  please let me know if it is ok

**Comment by kpedro88** [comment]
@cmsbuild please test

**Comment by calabria** [comment]
@kpedro @HuguesBrun please check the changes to HcalDetIdAssociator

**Comment by kpedro88** [comment]
@calabria that looks like it should help mitigate the problem as I described, we'll see what igprof says
(also, wrong tag in your post...)

**Comment by kpedro88** [comment]
@cmsbuild please test

**Comment by kpedro88** [comment]
Looks like HLT doesn't like that parameter addition (unsurprising in retrospect).

It might be better to add a generic `isPhase2` parameter to all `DetIdAssociatorESProducer` configs and the base class, and then specialize the Hcal class to return 1 for `getNumberOfSubdetectors()` if `isPhase2` is true. You may then have to do something with `HLTrigger/Configuration/python/customizeHLTforCMSSW.py` to add the parameter to the auto-generated HLT configs.

**Comment by calabria** [comment]
@kpedro88 I modified the customiHLT... and it seems everything works now. Do you think we really need to add a boolean to the base class?

**Comment by slava77** [comment]
@cmsbuild please test

File: HLTrigger/Configuration/python/customizeHLTforCMSSW.py [hunk 1]
```diff
@@ -58,6 +58,13 @@ def customiseFor14833(process):
                 producer.includeME0 = cms.bool(False)
     return process
 
+def customiseFor16670(process):
+    for producer in esproducers_by_type(process, "DetIdAssociatorESProducer"):
+        if (producer.ComponentName.value() == 'HcalDetIdAssociator'):
+            if not hasattr(producer,'hcalRegion'):
+                producer.hcalRegion = cms.int32(1)
```
Comment (kpedro88) [inline HLTrigger/Configuration/python/customizeHLTforCMSSW.py]: @calabria this looks okay, but now you've set all HLT versions of this module to ignore HE. The value should probably be set to 2 by default, though I wonder if we ever run this in HLT for Phase2.

Comment (calabria) [inline HLTrigger/Configuration/python/customizeHLTforCMSSW.py]: yes you are right

**Comment by kpedro88** [comment]
please abort

**Comment by kpedro88** [comment]
please test

**Comment by mdhildreth** [comment]
+1

**Comment by slava77** [comment]
looking at run-2 validation, there are changes in muon energy 
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_8_1_X_2016-11-21-1100+16670/17077/validateJR/all_OldVSNew_RunSinglePh2016Bwf136p731/all_OldVSNew_RunSinglePh2016Bwf136p731c_recoMuons_muons__reRECO_obj_calEnergy_hadS9.png

File: TrackingTools/TrackAssociator/plugins/HcalDetIdAssociator.h [hunk 2]
```diff
@@ -24,13 +24,16 @@ class HcalDetIdAssociator: public CaloDetIdAssociator{
...
+    {};
    
    virtual const char* name() const override { return "HCAL"; }
 
  protected:
-
-   virtual const unsigned int getNumberOfSubdetectors() const override { return 2;}
+    
+   int hcalReg_;
+   virtual const unsigned int getNumberOfSubdetectors(int hcalReg_) const { return hcalReg_;}
```
Comment (slava77) [inline TrackingTools/TrackAssociator/plugins/HcalDetIdAssociator.h]: this should be ```getNumberOfSubdetectors() ```

Comment (kpedro88) [inline TrackingTools/TrackAssociator/plugins/HcalDetIdAssociator.h]: Yes, that seems to be the problem. (The `override` can also be restored in that case.)

**Comment by calabria** [comment]
fixed, thanks!

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by civanch** [comment]
+1

**Comment by dmitrijus** [comment]
+1

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/16670 5029b88
- code changes are OK: to reduce CPU use in validation for run3 and 2023 setups and and in muon track-calorimeter association for 2023
- jenkins tests pass and comparisons with baseline show differences only in 2023 workflows in muon calorimeter/isolation quantities

Local test in CMSSW_8_1_X_2016-11-10-1100 of D4 21224 equivalent (ttbar) with PU200 shows expected reduction in CPU (in units of time corresponding to ~10 HS06 machine):
- about 800 s/evt in removed validation modules
- about 200 s/evt in remaining validation modules from a combination of improved code in MuonAssociatorByHitsHelper and from reduced eta range
- about 300 s/evt in muons1stStep (time changes:  ``` 326 s/ev ->     21 s/ev ``` )

There is also about 0.3 GiB/thread reduction in RSS memory.

**Comment by slava77** [comment]
backport of #16669

**Comment by Martin-Grunewald** [comment]
+1

**Comment by davidlange6** [comment]
+1

**Pull request merged**